7.0: add compute date

lepistone opened the pull request:
This module allows to recompute the delivery dates of outgoing moves based on when we expect to receive the goods.

The logic is different for "Make To Order" and "Make To Stock" products and is described in the **openerp**.py.

In the yaml tests you can find three use cases.

Thanks!

lepistone committed:
WIP initial tests for compute_delivery_date

lepistone committed:
add explicit dependency on sale_stock

I need this to be able to use the security_lead field in the company. It
is introduced by sale_stock

lepistone committed:
more WIP: we have an assertion

lepistone committed:
fix: write date on the lines

lepistone committed:
WIP production code for delivery dates

lepistone committed:
first complete test case passes

lepistone committed:
rename test file

lepistone committed:
always add the security days

lepistone committed:
add a second test case

lepistone committed:
add wizard and refactor to compute only on MTS

MTO logic is different and TODO. For the moment, we raise if we find a
MTO product

lepistone committed:
fix: search products from outgoing pickings

lepistone committed:
lint yaml

lepistone committed:
reduce logging to debug: it is normal behaviour

lepistone committed:
add mto logic, with a test

lepistone committed:
rename mts test cases now we have mto

lepistone committed:
provide a sample cron (disabled)

lepistone committed:
doc

lepistone committed:
add wizard for all products

lepistone committed:
split wizards in separate files

lepistone committed:
lint doc

lepistone committed:
use product's company, not user's

lepistone committed:
hide global wizard from "more" menu

lepistone committed:
test the global wizard

lepistone committed:
test also the wizard by product

lepistone committed:
lint

lepistone committed:
stock_user is enough, no need for managers here

jgrandguillaume commented:
Hi Leonardo,

I just finished to test this new module and it's just working like a charm ! 

Code review + test ok for me ! Thanks for the good job.

:+1:

lepistone committed:
fix: protect from None if security_days is zero

guewen commented:
Seems that running several times without new 'in' moves may change the dates, just need to investigate this further.

lepistone committed:
test twice and check idempotency

guewen commented on stock_picking_compute_delivery_date/wizard/all_products.xml:
```diff
@@ -0,0 +1,34 @@
+      view_mode="form"
+      target="new"
+      groups="base.group_no_one"
+      id="product_compute_all_delivery_dates_action"/>
```
This is misleading because it appears in 'Other options' on the products but will run on all the products, not the selected one. At least it appears only for the technical group. Is there a reason to have the entry in "Other Options" or could we just keep the menuitem.

guewen commented:
When a part of the moves fall in the PlanFinished trap, they keep their current 'date_expected'. I know this was on purpose, leaving room to implement that later, but I think that we should implement it now otherwise the computation will be biased: as the moves are sorted by `date_expected`, the ones that were in PlanFinished will be the first to be included the next time we run the computation, hence they will gain priority over the others and the others will be delayed. I think that we should write the last date of the plan (or last date of the plan + 1) on the moves that have not enough stock to be considered.
I will write a test and make a proposal.

lepistone commented:
@guewen +1

lepistone committed:
commit after comping dates for a product

Computing delivery dates for all products can take a few seconds per
product, and if there are many, they all go in a single long-lived
transaction that can be easily interrupted.

To avoid that, I create a new cursor and commit after dealing with every
product.

That feature can be disabled in situations where multiple cursors are
harmful, like when running the tests.

lepistone committed:
lint tests

jgrandguillaume commented:
Hi,

Just want to put my input about last @guewen comments:
- First you're complitely right, it doesn't work now as we expect this to work
- My idea at first was to order the moves not by expected date, but by creation date of the picking

Want we want here is : First arrived, first served.

So my suggestion would be that we order the moves by creation_date as it represent who should be delivered first.

This is an easy way to solve all this what do you think ?

Regards,

Joël

lepistone committed:
in mts, sort outgoing moves by creation date

This change should avoid circular situations where the plan finishes and
some outgoing moves are untouched, changing the order the next time that
the dates are recomputed. In that case the operation was not atomic,
something we would like to avoid.

Now the moves are sorted by creation date (which does not change) and at
the end the expected date is written. That way we should get always the
same results the next time.

lepistone commented:
I like your idea @jgrandguillaume, I did that. Thanks!

jgrandguillaume commented:
Looks good now ! Thanks for the fix. :+1:

guewen commented:
Using the `date` is fine because then the order will stay the same each time we launch the computation.
:+1:
